File: programmers/P43236.java
package programmers;

import java.util.Arrays;

public class P43236 {
    public int solution(int d, int[] rocks, int n) {
        Arrays.sort(rocks);
        int l = 0, h = d;
        while (l + 1 < h) {
            int dist = (l + h) / 2;
            int removed = 0, prev = 0;
            for (int rock : rocks) {
                if (rock - prev < dist) {
                    if (++removed > n) break;
                } else {
                    prev = rock;
                }
            }
            removed += d - prev < dist ? 1 : 0;
            if (removed > n) h = dist;
            else l = dist;
        }
        return l;
    }

    public static void main(String[] args) {
        System.out.println(new P43236().solution(25, new int[] { 2, 4, 11, 21, 17 }, 2));
    }
}

File: programmers/P43236.py
from time import perf_counter_ns as ns


def solution(d, rocks, n):
    stones = [0] + sorted(rocks) + [d]
    low, high = 0, d
    while low + 1 < high:
        dist = (low + high) // 2
        removed = failed = prev = 0
        for s in stones[1:]:
            if s - prev < dist:
                if removed == n:
                    failed = 1
                    break
                removed += 1
                continue
            prev = s
        if not failed:
            low = dist
        else:
            high = dist
    return low


if __name__ == '__main__':
    ITERATIONS = 1_000
    print(f'Running the basic tests {ITERATIONS:,} times...')
    tests = (
        (25, [2, 14, 11, 21, 17], 2, 4),
    )
    for distance, rocks, num_removal, expected in tests:
        print(f'solution({distance}, {rocks}, {num_removal}) returned', end=' ')
        if (result := solution(distance, rocks, num_removal)) == expected:
            print(f'the expected result {expected}', end=' ')
            fastest = float('inf')
            slowest = total = 0
            for _ in range(ITERATIONS):
                start = ns()
                solution(distance, rocks, num_removal)
                end = ns()
                time = end - start
                fastest, slowest = min(time, fastest), max(time, slowest)
                total += time
            print(f'in an average of {total / ITERATIONS / 1e3:,.2f}μs '
                  f'(min: {fastest / 1e3:,.2f}μs, '
                  f'max: {slowest / 1e3:,.2f}μs)')
        else:
            print(f'a wrong result {result} (expected: {expected})')
